src/shogun/preprocessor/RandomFourierGaussPreproc.cpp

```diff
@@ -7,394 +7,87 @@
-	dim_feature_space = dim_feature_space2;
-	dim_input_space = dim_input_space2;
-	kernelwidth=kernelwidth2;
-
-	SG_FREE(randomcoeff_multiplicative);
-	randomcoeff_multiplicative=NULL;
-	SG_FREE(randomcoeff_additive);
-	randomcoeff_additive=NULL;
-
-	cur_dim_feature_space=dim_feature_space;
-	cur_dim_input_space = dim_input_space;
-	cur_kernelwidth=kernelwidth;
-
-	if( (dim_feature_space>0) && (dim_input_space>0) )
-	{
-	randomcoeff_additive=SG_MALLOC(float64_t, cur_dim_feature_space);
-	randomcoeff_multiplicative=SG_MALLOC(float64_t, cur_dim_feature_space*cur_dim_input_space);
-
-	std::copy(randomcoeff_additive2, randomcoeff_additive2
-			+ dim_feature_space, randomcoeff_additive);
-	std::copy(randomcoeff_multiplicative2, randomcoeff_multiplicative2
-			+ cur_dim_feature_space*cur_dim_input_space, randomcoeff_multiplicative);
-	}
+	m_basis = sample_spectral_density(dim_input_space);
+	m_offset = SGVector<float64_t>(m_dim_output);
 
+	UniformRealDistribution<float64_t> uniform(0.0, 2.0 * Math::PI);
+	random::fill_array(m_offset, uniform, m_prng);
 }
 
 void RandomFourierGaussPreproc::fit(std::shared_ptr<Features> f)
 {
-	if (dim_feature_space <= 0) {
-		throw ShogunException(
-				"CRandomFourierGaussPreproc::init (Features *f): dim_feature_space<=0 is not allowed, use void set_dim_feature_space(const int32 dim) before!\n");
-	}
-
-	io::info("calling CRandomFourierGaussPreproc::init(...)");
-	int32_t num_features =
-	    f->as<DenseFeatures<float64_t>>()->get_num_features();
-
-	if (!test_rfinited()) {
-		dim_input_space = num_features;
-		init_randomcoefficients();
-		ASSERT( test_rfinited())
-	} else {
-		dim_input_space = num_features;
-		// does not reinit if dimension is the same to avoid overriding a previous call of set_randomcoefficients(...)
-		init_randomcoefficients();
-	}
+	auto num_features = f->as<DenseFeatures<float64_t>>()->get_num_features();
+	require(num_features > 0, "Input space dimension must be greater than zero");
+	
+	init_basis(num_features);
+	m_fitted = true;
 }
 
 SGVector<float64_t> RandomFourierGaussPreproc::apply_to_feature_vector(SGVector<float64_t> vector)
 {
-	if (!test_rfinited()) {
-		throw ShogunException(
-				"float64_t * CRandomFourierGaussPreproc::apply_to_feature_vector(...): test_rfinited()==false: you need to call before CRandomFourierGaussPreproc::init (Features *f) OR	1. set_dim_feature_space(const int32 dim), 2. set_dim_input_space(const int32 dim), 3. init_randomcoefficients() or set_randomcoefficients(...) \n");
-	}
-
-	float64_t val = std::sqrt(2.0 / cur_dim_feature_space);
-	SGVector<float64_t> res(cur_dim_feature_space);
-
-	for (int32_t od = 0; od < cur_dim_feature_space; ++od) {
-		SGVector<float64_t> wrapper(randomcoeff_multiplicative+od*cur_dim_input_space, cur_dim_input_space, false);
-		res[od] = val * cos(randomcoeff_additive[od] + linalg::dot(vector, wrapper));
-	}
-
-	return res;
+	return static_cast<SGVector<float64_t>>(
+	    apply_to_matrix(static_cast<SGMatrix<float64_t>>(vector)));
```

Comment:
does this actually work and is safe?
@gf712 I didnt know we can just cast these types back and forth ....

Reply:
especially the cast from SGMatrix->SGVector ... what if there are more than one column in the matrix (it wont be here) is this checked somewhere?

Reply:
I thought doing this via a constructor would be better/safer (as said in irc @jonpsy )

Reply:
this will just call https://github.com/shogun-toolbox/shogun/blob/ab6a69686e751f819aec3adf6b9f2d7df1c2644c/src/shogun/lib/SGMatrix.h#L116

Reply:
and https://github.com/shogun-toolbox/shogun/blob/ab6a69686e751f819aec3adf6b9f2d7df1c2644c/src/shogun/lib/SGVector.h#L72

Reply:
https://github.com/shogun-toolbox/shogun/blob/develop/src/shogun/lib/SGMatrix.h#L116

Reply:
https://github.com/shogun-toolbox/shogun/blob/develop/src/shogun/lib/SGVector.h#L72

Reply:
I just learned that `static_cast` in fact calls the ctor. Nevermind then :)

Reply:
sorry my inet lagged and didnt see your reply. Cool!